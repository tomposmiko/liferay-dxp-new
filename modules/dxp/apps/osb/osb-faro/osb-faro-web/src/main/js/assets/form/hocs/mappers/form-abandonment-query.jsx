import React from 'react';
import {getPercentage} from 'shared/util/util';
import {getVariables, safeResultToProps} from 'shared/util/mappers';
import {sub} from 'shared/util/lang';

/**
 * Get Label Value
 * @param {number} value
 * @returns {string} label
 */
const getLabelHeader = value =>
	sub(Liferay.Language.get('x-form-steps'), [value]);

/**
 * Get Label Fields
 * @description if value is equal 1, returns label in the singular
 * @param {number} value
 * @returns {string} label
 */
const getLabelFields = value => {
	if (value === 1) {
		return sub(Liferay.Language.get('x-field'), [value]);
	}

	return sub(Liferay.Language.get('x-fields'), [value]);
};

/**
 * Get Color
 * @description if the index is even, return a color
 * if it is odd return another color
 * @param {number} index
 * @returns {string} color
 */
const getColor = index => (index % 2 === 0 ? '#B1D4FF' : '#95C5FF');

const mapResultToProps = safeResultToProps(({form}) => {
	const {formPageMetrics} = form;

	const header = [
		{
			icon: 'document',
			label: getLabelHeader(formPageMetrics.length)
		},
		{
			icon: 'custom-field',
			label: getLabelFields(
				formPageMetrics
					.map(({formFieldMetrics}) => formFieldMetrics.length)
					.reduce((total, num) => total + num, 0)
			)
		}
	];

	const items = formPageMetrics.map(
		(
			{
				formFieldMetrics,
				pageAbandonmentsMetric,
				pageName,
				pageViewsMetric
			},
			index
		) => ({
			columns: [
				{
					icon: 'document',
					label: () => (
						<>
							{sub(Liferay.Language.get('step-x'), [index + 1])}{' '}
							{pageName && <strong>{` ${pageName}`}</strong>}
						</>
					)
				},
				{
					icon: 'custom-field',
					label: getLabelFields(formFieldMetrics.length)
				}
			],
			expanded: index === 0 ? true : false,
			items: formFieldMetrics.map(
				({fieldAbandonmentsMetric, fieldName}, index) => ({
					columns: [
						{
							icon: 'custom-field',
							label: fieldName
						}
					],
					progress: [
						{
							color: getColor(index),
							value: `${getPercentage(
								fieldAbandonmentsMetric.value,
								pageViewsMetric.value
							)}%`
						}
					]
				})
			),
			progress: [
				{
					color: '#4B9BFF',
					value: `${getPercentage(
						pageAbandonmentsMetric.value,
						pageViewsMetric.value
					)}%`
				}
			],
			showControls: formPageMetrics.length === 1 ? false : true
		})
	);

	return {
		header,
		items
	};
});

/**
 * Map Props to Options
 * @param {object} param0 props
 * @param {object} param1 context
 */
const mapPropsToOptions = ({
	filters,
	interval,
	rangeSelectors,
	router: {params}
}) => getVariables({filters, interval, params, rangeSelectors});

export {mapPropsToOptions, mapResultToProps};
