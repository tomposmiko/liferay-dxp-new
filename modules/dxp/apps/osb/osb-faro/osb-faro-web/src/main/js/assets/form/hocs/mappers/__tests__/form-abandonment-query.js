import {mapPropsToOptions, mapResultToProps} from '../form-abandonment-query';

const router = {
	params: {
		assetId: 'formId',
		title: 'Liferay',
		touchpoint: 'Any'
	}
};

const data = {
	form: {
		formPageMetrics: [
			{
				formFieldMetrics: [
					{
						fieldAbandonmentsMetric: {
							value: 5
						},
						fieldEmptyMetric: {
							value: 0
						},
						fieldInteractionDurationMetric: {
							value: 6230
						},
						fieldInteractionsMetric: {
							value: 812
						},
						fieldName: 'Name',
						fieldRefilledMetric: {
							value: 8
						}
					},
					{
						fieldAbandonmentsMetric: {
							value: 16
						},
						fieldEmptyMetric: {
							value: 0
						},
						fieldInteractionDurationMetric: {
							value: 8410
						},
						fieldInteractionsMetric: {
							value: 796
						},
						fieldName: 'Address',
						fieldRefilledMetric: {
							value: 12
						}
					},
					{
						fieldAbandonmentsMetric: {
							value: 4
						},
						fieldEmptyMetric: {
							value: 0
						},
						fieldInteractionDurationMetric: {
							value: 2100
						},
						fieldInteractionsMetric: {
							value: 792
						},
						fieldName: 'City',
						fieldRefilledMetric: {
							value: 2
						}
					},
					{
						fieldAbandonmentsMetric: {
							value: 12
						},
						fieldEmptyMetric: {
							value: 0
						},
						fieldInteractionDurationMetric: {
							value: 1700
						},
						fieldInteractionsMetric: {
							value: 780
						},
						fieldName: 'State',
						fieldRefilledMetric: {
							value: 1.77
						}
					},
					{
						fieldAbandonmentsMetric: {
							value: 7
						},
						fieldEmptyMetric: {
							value: 0
						},
						fieldInteractionDurationMetric: {
							value: 2890
						},
						fieldInteractionsMetric: {
							value: 773
						},
						fieldName: 'Zip Code',
						fieldRefilledMetric: {
							value: 32
						}
					},
					{
						fieldAbandonmentsMetric: {
							value: 221
						},
						fieldEmptyMetric: {
							value: 0
						},
						fieldInteractionDurationMetric: {
							value: 2090
						},
						fieldInteractionsMetric: {
							value: 34
						},
						fieldName: 'Order Number',
						fieldRefilledMetric: {
							value: 23
						}
					},
					{
						fieldAbandonmentsMetric: {
							value: 3
						},
						fieldEmptyMetric: {
							value: 0
						},
						fieldInteractionDurationMetric: {
							value: 3900
						},
						fieldInteractionsMetric: {
							value: 766
						},
						fieldName: 'Phone Number',
						fieldRefilledMetric: {
							value: 11
						}
					},
					{
						fieldAbandonmentsMetric: {
							value: 19
						},
						fieldEmptyMetric: {
							value: 0
						},
						fieldInteractionDurationMetric: {
							value: 2870
						},
						fieldInteractionsMetric: {
							value: 670
						},
						fieldName: 'Requested Repair Date',
						fieldRefilledMetric: {
							value: 7.3
						}
					},
					{
						fieldAbandonmentsMetric: {
							value: 17
						},
						fieldEmptyMetric: {
							value: 0
						},
						fieldInteractionDurationMetric: {
							value: 62230
						},
						fieldInteractionsMetric: {
							value: 215
						},
						fieldName: 'Description of Problem',
						fieldRefilledMetric: {
							value: 28
						}
					}
				],
				pageAbandonmentsMetric: {
					value: 461
				},
				pageIndex: '0',
				pageViewsMetric: {
					value: 1273
				}
			}
		]
	}
};

describe('FormAbandonmentQuery mapper', () => {
	it('should extract items from result', () => {
		const props = mapResultToProps({data});

		expect(props).toHaveProperty('header');
		expect(props).toHaveProperty('items');
	});

	it('should return a object with empty true when formPageMetrics is not present in parameters', () => {
		const props = mapResultToProps({
			data: {
				form: {
					formPageMetrics: []
				}
			}
		});

		expect(props).toMatchSnapshot();
	});
	it('should include options', () => {
		const options = mapPropsToOptions({
			filters: {},
			rangeSelectors: {rangeKey: '7'},
			router
		});

		expect(options).toEqual({
			variables: {
				assetId: 'formId',
				devices: 'Any',
				location: 'Any',
				rangeEnd: null,
				rangeKey: 7,
				rangeStart: null,
				title: 'Liferay',
				touchpoint: null
			}
		});
	});
});
