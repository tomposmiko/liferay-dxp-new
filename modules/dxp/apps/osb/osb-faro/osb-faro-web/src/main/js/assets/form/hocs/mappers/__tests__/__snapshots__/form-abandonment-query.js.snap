// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormAbandonmentQuery mapper should return a object with empty true when formPageMetrics is not present in parameters 1`] = `
Object {
  "error": null,
  "header": Array [
    Object {
      "icon": "document",
      "label": "0 Form Steps",
    },
    Object {
      "icon": "custom-field",
      "label": "0 Fields",
    },
  ],
  "items": Array [],
  "loading": false,
  "refetch": undefined,
}
`;
